tests/test_fv3run.py

```diff
@@ -0,0 +1,243 @@
+
+def subprocess_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        subprocess.check_call(
+            ['fv3run', config_file.name, outdir, '--runfile', MOCK_RUNSCRIPT]
+        )
+
+
+def docker_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        try:
+            subprocess.check_call(
+                ['fv3run', config_file.name, outdir, '--runfile', MOCK_RUNSCRIPT,
+                 '--dockerimage', DOCKER_IMAGE_NAME]
+            )
+        except subprocess.CalledProcessError as err:
+            if os.path.isfile(os.path.join(outdir, 'stdout.log')):
+                with open(os.path.join(outdir, 'stdout.log')) as stdout:
+                    print(f'STDOUT: {stdout.read()}')
+            if os.path.isfile(os.path.join(outdir, 'stderr.log')):
+                with open(os.path.join(outdir, 'stderr.log')) as stderr:
+                    print(f'STDERR: {stderr.read()}')
+            raise err
+
+
+def config_dict_module_run(config_dict, outdir):
+    fv3config.run(config_dict, outdir, runfile=MOCK_RUNSCRIPT)
+
+
+def config_dict_filename_run(config_dict, outdir):
+    with tempfile.NamedTemporaryFile(mode='w') as config_file:
+        config_file.write(yaml.dump(config_dict))
+        config_file.flush()
+        fv3config.run(config_file.name, outdir, runfile=MOCK_RUNSCRIPT)
+
+
+def count_executed_ranks(rundir):
+    filenames = os.listdir(rundir)
+    total = 0
+    while f'rank_{total}' in filenames:
+        total += 1
+    return total
+
+
+@contextlib.contextmanager
+def mocked_check_call():
+    call_list = []
+
+    def mock_check_call(arg_list, *args, **kwargs):
+        call_list.append(arg_list)
+        if arg_list[0] == 'mpirun':
+            pass
+        else:
+            original_check_call(arg_list, *args, **kwargs)
+    subprocess.check_call = mock_check_call
+    try:
+        yield call_list
+    finally:
+        subprocess.check_call = original_check_call
+
+
+@contextlib.contextmanager
+def cleaned_up_directory(dirname):
+    try:
+        yield
+    finally:
+        if os.path.isdir(dirname):
+            shutil.rmtree(dirname)
+
+
+@pytest.mark.parametrize(
+    "runner",
+    [docker_run, config_dict_module_run, config_dict_filename_run]
+)
+def test_fv3run_without_mpi(runner):
+    # the test for docker_run *will* use MPI inside the docker container
+    # we're only making it so subprocess calls of mpirun don't execute
+    if runner == docker_run and not DOCKER_ENABLED:
+        pytest.skip(f'docker or docker image {DOCKER_IMAGE_NAME} is not available')
+    fv3config.ensure_data_is_downloaded()
+    outdir = os.path.join(TEST_DIR, 'outdir')
+    with mocked_check_call(), cleaned_up_directory(outdir):
+        runner(fv3config.get_default_config(), outdir)
```

Comment:
There are no assert statements in this test. What exactly is it testing?

Reply:
Short answer: not much! These tests are a bit of a mess. Overhauling now.